src/backend/distributed/executor/multi_client_executor.c

```diff
@@ -653,6 +661,155 @@ MultiClientCopyData(int32 connectionId, int32 fileDescriptor)
 
 
 /*
+ * MultiClientCreateWaitInfo creates a WaitInfo structure, capable of keeping
+ * track of what maxConnections connections are waiting for; to allow
+ * efficiently waiting for all of them at once.
+ *
+ * Connections can be added using MultiClientAddWait(). All added connections
+ * can then be waited upon together using MultiClientWait().
+ */
+WaitInfo *
+MultiClientCreateWaitInfo(int maxConnections)
+{
+	WaitInfo *waitInfo = palloc(sizeof(WaitInfo));
+
+	waitInfo->allocated = maxConnections;
+	waitInfo->waiting = 0;
+	waitInfo->pollfds = palloc(maxConnections * sizeof(struct pollfd));
+
+	return waitInfo;
+}
+
+
+/* MultiClientResetWaitInfo clears all pending waits from a WaitInfo. */
+void
+MultiClientResetWaitInfo(WaitInfo *waitInfo)
+{
+	waitInfo->waiting = 0;
+	waitInfo->haveBusy = false;
+	waitInfo->haveFailure = false;
+}
+
+
+/* MultiClientFreeWaitInfo frees a resources associated with a waitInfo struct. */
+void
+MultiClientFreeWaitInfo(WaitInfo *waitInfo)
+{
+	pfree(waitInfo->pollfds);
+	pfree(waitInfo);
+}
+
+
+/*
+ * MultiClientWaitFor adds a connection to be waited upon, waiting for
+ * waitStatus.
+ */
+void
+MultiClientRememberWait(WaitInfo *waitInfo, WaitStatus waitStatus, int32 connectionId)
```

Comment:
What do you think about `MultiClientRegisterWait`? I noticed in another comment you referenced `MultiClientAddWait`. Is there a reason that name was abandoned? `Remember` is a weird way of saying "add to array".